### DISCUSSION TITLE
Use getStaticProps and getServerSideProps together ?

### DISCUSSION THREAD

**Discussion by fimbault** [vercel/next.js#19606] [Help]
Hi, 

We're using nextjs 10. We'd like to find a way to be able to do both
- build time data fetching (for simple config, such as internationalization or similar local static requirements)
- and runtime rendering (for more complex services hosted elsewhere that really need to be dynamic)

Our typical use case is a blog page which embeds a search engine. But the need is very generic.

Initially, we had been using nextjs with prisma (or even a simple json) to store some content, works fine using getStaticProps.
Now we're trying to also call a search engine (in our case meilisearch) via HTTP calls, so we moved everything (including the part on prisma) to getServerSideProps. Locally it works fine. 

But when we deploy on vercel, we get an issue during deployment, probably because it doesn't limit prisma to read-only access as was previously the case with getStaticProps. 

with prisma, it fails
![image](https://user-images.githubusercontent.com/45664395/100469846-59586800-30d7-11eb-882e-373966b3bbb2.png)

same with a json config file
![image (2)](https://user-images.githubusercontent.com/45664395/100472240-282e6680-30dc-11eb-88d2-6351525f7a1a.png)

It doesn't seem that we can use both getStaticProps (for content in prisma or json) and getServerSideProps (for meilisearch) in the same page. 

It seems there are old discussions and issues related to this problem, but I haven't seen a clear answer.  
https://github.com/vercel/next.js/discussions/11424
https://github.com/vercel/next.js/issues/7208

What is the suggested path forward? or workaround?
Thanks.

**Comment by lxsmnsyc** [comment]
You can only use one of the two on a single page, but not both

**Reply by fimbault** [reply]
Thks for your feedback, but this is not answering my question (how can we go around it because there are valid use cases where it matters). I've seen indeed that both are mutually exclusive, but the problem is that getServerSideProps limits a lot what one can do. Typically file access won't work (browser and node environments being quite different, obviously). 

The problem overall is that getServerSideProps and getStaticProps correspond to very different scenarios, so not being able to do both is a real pain.

But maybe people have come around that with specific architectures, like not putting everything into the page and calling external services from a sub-component ? Any hint would be welcome.

Of course one option is to get everything to work as external webservices, but then to be honest, why would I use nextjs at all? Our objective would be to use static when possible, and serverless functions when we really need something dynamic.

**Reply by mAAdhaTTah** [reply]
> like not putting everything into the page and calling external services from a sub-component ?

This is what first came to mind when you said you had a search piece. If you used Next.js' API endpoints, you could hit them from the client-side without needing to include that info in gSSP.

> The problem overall is that getServerSideProps and getStaticProps correspond to very different scenarios, so not being able to do both is a real pain.

I'm honestly not sure what you get out of using both, besides perhaps a built-in caching layer. Once you need dynamic props, the whole page needs to be dynamically rendered; you don't get the static generation of that page you would get w/ gSP. If the goal is that caching, you're unfortunately stuck doing it manually (fs, redis, memcached, etc.) cuz the purpose of those functions from Next.js' perspective isn't caching, it's static generation, which you're opting out of by using gSSP.

**Comment by iliketomatoes** [comment]
We're facing the same issue at the company where I work.
On a specific page, we're trying to fetch data from Contentful at build time through `getStaticProps`, and then get dynamic data for each request through `getServerSideProps`. 

Just wanted to share my experience here, because I think there are several users out there who'd like to leverage a similar feature.

**Reply by lxsmnsyc** [reply]
There could be two solutions:
1. getStaticProps with revalidate set to true.
2. getStaticProps intermingled with data-fetching libraries (e.g. SWR) and a data preloader (link rel="preload")

No. 2 isn't SEO-friendly. And again, you can't mix `getServerSideProps` and `getStaticProps` since those two behave entirely different from one another.

**Reply by fimbault** [reply]
Thanks a lot will test and let you know.

**Reply by AlexSapoznikov** [reply]
I am currently exactly in the same position. How did you solve this?

**Reply by fimbault** [reply]
well I didn't per say, The suggested workaround didn't work for our use case. 
We had to remove some of the things we wanted because getServerSideProps was something important for that specific usecase, and so for that piece of software we couldn't provide static support for the pieces that would have benefited from it. We can live with that for the time being, but clearly being able to choose how you mix the 2 behaviors would be a major breakthrough.

**Reply by AlexSapoznikov** [reply]
Thanks for reply.

**Comment by masterkingmk** [comment]
I dont good developer my friend but I have a lot idea and like find good dev to create 6 crypto coin another.if you like help me I very happy.


Morteza kargozar Managment of Iran Apple store

On Thursday, January 28, 2021, 03:25, Alex <notifications@github.com> wrote:




I am currently exactly in the same position. How did you solve this?
